dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -221,249 +272,371 @@ namespace winrt::Microsoft::Windows::PushNotifications::implementation
+                RegisterSinkHelper();
                 {
-                    auto lock{ s_activatorInfoLock.lock_shared() };
-                    THROW_HR_IF_MSG(E_INVALIDARG, s_comActivatorRegistration, "ComActivator already registered.");
+                    auto lock{ m_lock.lock_exclusive() };
+                    m_singletonForegroundRegistration = true;
+                    m_singletonBackgroundRegistration = true;
                 }
 
-                GetWaitHandleForArgs().create();
-                GetComServerClsid() = taskClsid;
-                
-                THROW_IF_FAILED(::CoRegisterClassObject(
-                    taskClsid,
-                    winrt::make<PushNotificationBackgroundTaskFactory>().get(),
-                    CLSCTX_LOCAL_SERVER,
-                    REGCLS_MULTIPLEUSE,
-                    &s_comActivatorRegistration));
-            }
+                // Register a COM object for the PushNotificationLongRunningProcess to CoCreate in background activation scenarios
+                if (AppModel::Identity::IsPackagedProcess())
+                {
+                    auto lock{ m_lock.lock_exclusive() };
+                    THROW_IF_FAILED(::CoRegisterClassObject(
+                        m_registeredClsid,
+                        winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                        CLSCTX_LOCAL_SERVER,
+                        REGCLS_MULTIPLEUSE,
+                        &m_comActivatorRegistration));
+                }
 
-            if (builder)
-            {
-                registeredTaskFromBuilder = builder.Register();
+                scopeExitToCleanRegistrations.release();
             }
-
-            scopeExitToCleanRegistrations.release();
-            auto lock{ s_activatorInfoLock.lock_exclusive() };
-            s_pushTriggerRegistration = registeredTaskFromBuilder;
-            PushNotificationTelemetry::ActivatorRegisteredByApi(S_OK, details.Activators());
         }
-
-        catch(...)
+        catch (...)
         {
-            PushNotificationTelemetry::ActivatorRegisteredByApi(wil::ResultFromCaughtException(),
-                details == nullptr ? PushNotificationRegistrationActivators::Undefined : details.Activators());
+            PushNotificationTelemetry::ActivatorRegisteredByApi(wil::ResultFromCaughtException());
             throw;
         }
+
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            m_registering = false;
+        }
     }
 
-    void PushNotificationManager::UnregisterActivator(PushNotificationRegistrationActivators const& activators)
+    void PushNotificationManager::Unregister()
     {
-        THROW_HR_IF(E_NOTIMPL, !::Microsoft::Windows::PushNotifications::Feature_PushNotifications::IsEnabled());
+        {
+            auto lock{ m_lock.lock_exclusive() };
+            THROW_HR_IF_MSG(E_FAIL, m_registering, "Register or Unregister currently in progress!");
```

Comment:
rename m_registering?

Reply:
I think in the meeting we decided it is fine for now. :)